linux/hid.c

```diff
@@ -493,127 +495,229 @@ parse_uevent_info(const char *uevent, unsigned *bus_type,
 }
 
 
-static int get_device_string(hid_device *dev, enum device_string_id key, wchar_t *string, size_t maxlen)
+static void get_device_info_for_device(struct udev_device *raw_dev, unsigned short vendor_id, unsigned short product_id, struct hid_device_info ** first_dev, struct hid_device_info ** last_dev)
 {
-	struct udev *udev;
-	struct udev_device *udev_dev, *parent, *hid_dev;
-	struct stat s;
-	int ret = -1;
+	struct hid_device_info *cur_dev = NULL;
+	struct hid_device_info *prev_dev = NULL; /* previous device */
+
+	const char *sysfs_path;
+	const char *dev_path;
+	const char *str;
+	struct udev_device *hid_dev; /* The device's HID udev node. */
+	struct udev_device *usb_dev; /* The device's USB udev node. */
+	struct udev_device *intf_dev; /* The device's interface (in the USB sense). */
+	unsigned short dev_vid;
+	unsigned short dev_pid;
 	char *serial_number_utf8 = NULL;
 	char *product_name_utf8 = NULL;
+	unsigned bus_type;
+	int result;
+	struct hidraw_report_descriptor report_desc;
 
-	if (!string || !maxlen) {
-		register_device_error(dev, "Zero buffer/length");
-		return -1;
+	sysfs_path = udev_device_get_syspath(raw_dev);
+	dev_path = udev_device_get_devnode(raw_dev);
+
+	hid_dev = udev_device_get_parent_with_subsystem_devtype(
+		raw_dev,
+		"hid",
+		NULL);
+
+	if (!hid_dev) {
+		/* Unable to find parent hid device. */
+		goto next;
+	}
+
+	result = parse_uevent_info(
+		udev_device_get_sysattr_value(hid_dev, "uevent"),
+		&bus_type,
+		&dev_vid,
+		&dev_pid,
+		&serial_number_utf8,
+		&product_name_utf8);
+
+	if (!result) {
+		/* parse_uevent_info() failed for at least one field. */
+		goto next;
+	}
+
+	/* Filter out unhandled devices right away */
+	switch (bus_type) {
+		case BUS_BLUETOOTH:
+		case BUS_I2C:
+		case BUS_USB:
+			break;
+
+		default:
+			goto next;
+	}
+
+	/* Check the VID/PID against the arguments */
+	if ((vendor_id == 0x0 || vendor_id == dev_vid) &&
+		(product_id == 0x0 || product_id == dev_pid)) {
```

Comment:
this is filtering logic - should be a part of `hid_enumerate` only, and not a part of `get_device_info_for_device`

Reply:
To get the vendor_id and product_id from `raw_dev`, it requires 2 calls to udev methods, and some string parsing (`parse_uevent_info`). 
Having to repeat that much logic doesnt feel good, but I do appreciate that this is also not great